### ISSUE TITLE
add  external ram  support     on  basys3 Implementation

### ISSUE THREAD

**Issue by zzsczz** [wfjm/w11#39]
Onboard   memory  size of 176 kB  is too small to boot  211bsd   .

https://github.com/machdyne/qqspi

provides  design of ram  with Pmod™ compatible interface which can be mounted on  basys3 .

The ram chip  [APS6404L-3SQR]  manufacture[apmemory] provide the  verilog model of the chip through emal.
https://www.apmemory.com/products/psram-iot-ram/

Can w11 add support for basys3  + pmod sram configuration  ?

**Comment by wfjm** [comment]
Hello @zzsczz,  
thanks for the interest in the w11 project. The [machdyne/qqspi](https://github.com/machdyne/qqspi) is a very interesting product indeed. And in principle it is possible to add a memory controller for this type of memory.
This is not only interesting for the Basys3, but also for the CmodA7. The Basys3 is quite expensive, and not the best choice for a w11 system (buy an ArtyA7-35 for similar cost). I made a Basys3 design (see [link](https://github.com/wfjm/w11/tree/master/rtl/sys_gen/w11a/basys3)) because I have one for other reasons. The CmodA7-35T is the lowest cost board currently supported in w11 (see [link](https://github.com/wfjm/w11/tree/master/rtl/sys_gen/w11a/cmoda7)). And with 672 kB a bit short of memory, so the QSPI memory is an interesting option. So overall an interesting option.

As usual, the devil is in the detail:
- the available verilog code can't be used because w11 is a pure VHDL project and will stay this way because whole verification is based on [GHDL](https://github.com/ghdl/ghdl).
- the physical memory interface is 32 bit wide with byte enables. The QSPI memory chips support byte write, but no bursts with byte enables. So the byte enable functionality must be part of the memory controller. Conceptually straight forward, but also some work.
- the read latency is 22 cycles. No problem, that will be mostly hidden by the cache.
- the write latency is 24-28 cycles. This is a problem because the cache is currently write-through and can create in some workloads significant memory write rates. The w11 SDRAM memory interface has therefore a write queue and write combining to cope with the also quite significant latencies of the SDRAM controller.

Bottom line is, that the controller will be quite complex, and inherit some features from the SDRAM controller.
Doable, very interesting, but it will not happen on a short time scale.

**Comment by zzsczz** [comment]
I get it.
I have a   nexys3(spartan-6) .  I will install  old ISE  dev tools to  build w11 and study .
Thank   @wfjm .

BTW  do you know any thing about  implementation vax on fpga?

**Comment by wfjm** [comment]
@zzsczz  wrote
> BTW do you know any thing about implementation vax on fpga?

I'm not aware of an open source implementation.
